### PR TITLE
Custom path to Optional class for Optional emptiness handler

### PR DESCRIPTION
**Pull request by shas19**
Allowing custom path to the Optional class with `-XepOpt:NullAway: CheckOptionalEmptinessCustomClasses =` flag.
This custom path is used by optional emptiness handler.  

No option at all passed in the command line => set is (java.utils.Optional)
Explicit option passed => the classes in that set + `java.utils.Optional`

 Added unit test

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java [hunk 1]
```diff
@@ -70,6 +70,8 @@
 
   protected boolean checkOptionalEmptiness;
 
+  protected String optionalClassPath;
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java]: Should probably be an `ImmutableSet<String>`, like with `unannotatedClasses`. It should also be `@Nullable` (for the case where `checkOptionalEmptiness` is `false`). We want to support the case of: a) No Optional classes passed, b) One Optional class passed (or taken from the default), c) N Optional classes passed. This field would only support (b)

File: nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java [hunk 2]
```diff
@@ -54,7 +55,11 @@
 
   @Override
   public void onMatchTopLevelClass(
-      NullAway analysis, ClassTree tree, VisitorState state, Symbol.ClassSymbol classSymbol) {
+      NullAway analysis,
+      ClassTree tree,
+      VisitorState state,
+      Symbol.ClassSymbol classSymbol,
+      Config config) {
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java]: Don't pass this here, pass it to the handler's constructor in `Handlers.buildDefault(...)`, that's the convention used by existing handlers.

File: nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java [hunk 3]
```diff
@@ -65,10 +64,14 @@ public boolean onOverrideMayBeNullExpr(
...
-      NullAway analysis, ClassTree tree, VisitorState state, Symbol.ClassSymbol classSymbol) {
+      NullAway analysis,
+      ClassTree tree,
+      VisitorState state,
+      Symbol.ClassSymbol classSymbol,
+      Config config) {
     if (optionalType == null) {
       optionalType =
-          Optional.ofNullable(state.getTypeFromString(OPTIONAL_PATH))
+          Optional.ofNullable(state.getTypeFromString(config.getOptionalClassPath()))
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: Must be generalized to deal with multiple `Optional` classes.

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 4]
```diff
@@ -1233,6 +1233,65 @@ public void OptionalEmptinessHandlerTest() {
...
+  public void OptionalEmptinessHandlerWithCustomPathTest() {
+    compilationHelper
+        .setArgs(
+            Arrays.asList(
+                "-d",
+                temporaryFolder.getRoot().getAbsolutePath(),
+                "-XepOpt:NullAway:AnnotatedPackages=com.uber",
+                "-XepOpt:NullAway:UnannotatedSubPackages=com.uber.lib.unannotated",
+                "-XepOpt:NullAway:CheckOptionalEmptiness=true",
+                "-XepOpt:NullAway:OptionalClassPath=com.google.common.base.Optional"))
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: :+1: Would suggest also:

1) Testing for `"-XepOpt:NullAway:OptionalClassPath=\"\""` (I'd say this should just raise a config error if `XepOpt:NullAway:CheckOptionalEmptiness` is set)
2) Testing for `"-XepOpt:NullAway:OptionalClassPath=java.lang.Optional,com.google.common.base.Optional"`

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Should we not have `java.utils.Optional` by default in the set?

Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: The way I would do it, if possible:

1) No option at all passed in the command line => set is (`java.utils.Optional`)
2) Explicit option passed, empty or unparsable => error
3) Explicit option passed, parsable non-empty set => the classes in that set, and nothing else

This allows people using this handler for their own custom `Optional` without enabling it for `java.utils.Optional`. Then again, not sure if it's more consistent with how we do other stuff to just have  `java.utils.Optional` always on (when `CheckOptionalEmptiness` is set), then rename this option to something like `CustomOptionalClasses`. Maybe on second thought, I am liking this later way of doing it better.

@msridhar , thoughts ?

Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: I think we should always check `java.util.Optional` and this option should add additional aliases for it.

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Okay then. So first two cases as Lazaro described and and in case #3 Explicit option passed, parsable non-empty set => the classes in that set + `java.utils.Optional`. 
That works @lazaroclapp?

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: I am not sure if we really need the unparsable check. Since we do not have it for other configs. Also I'm not sure how to define it exactly. So let me know if you think it should be necessary. 
@msridhar @lazaroclapp

File: nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java (hunk 2 above)
Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java]: Before I go ahead and do a full pass, it looks to me like this remains to be addressed :)

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 4 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: I think given what Manu is saying, then there is just two cases:

1. No option at all passed in the command line => set is (java.utils.Optional)
2. `CustomOptionalClasses=a.b.c.A1;a.b.c.A2` => set is (java.utils.Optional,a.b.c.A1,a.b.c.A2)

Still worth to check that it works with zero (no `CustomOptionalClasses` option), one and two classes being passed.

File: nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java [hunk 5]
```diff
@@ -56,7 +56,7 @@
   static final String FL_ACKNOWLEDGE_RESTRICTIVE =
       EP_FL_NAMESPACE + ":AcknowledgeRestrictiveAnnotations";
   static final String FL_CHECK_OPTIONAL_EMPTINESS = EP_FL_NAMESPACE + ":CheckOptionalEmptiness";
-  static final String FL_OPTIONAL_CLASS_PACKAGE = EP_FL_NAMESPACE + ":OptionalClassPath";
+  static final String FL_OPTIONAL_CLASS_PATHS = EP_FL_NAMESPACE + ":OptionalClassPaths";
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java]: I still think this should be `CustomOptionalClasses`, otherwise I keep parsing it on my head as "Optional (class paths)" rather than "(Optional class) paths"

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java]: Not that my alternative is perfect, but... mmm... maybe `CheckOptionalEmptinessCustomClasses`? A bit verbose, but at least ties the two options together.

Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/ErrorProneCLIFlagsConfig.java]: Changed it to `CheckOptionalEmptinessCustomClasses`, sounds better

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 4 above)
Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Added test for zero, one and two custom paths case.

File: nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java (hunk 2 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/ApacheThriftIsSetHandler.java]: Done now 😅

File: nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java (hunk 1 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/AbstractConfig.java]: Done 👍

File: nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java [hunk 6]
```diff
@@ -49,9 +51,12 @@
  */
 public class OptionalEmptinessHandler extends BaseNoOpHandler {
 
-  private static String OPTIONAL_PATH = "java.util.Optional";
+  @Nullable private Set<Optional<Type>> optionalTypes;
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: Why is this a `Set<Optional<Type>>` rather than a `Set<Type>`? We could just filter out any type paths for which `state.getTypeFromString(type)` returns `null`.

Also, it probably should be an `ImmutableSet`, since it won't change after the call to `onMatchTopLevelClass`

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java [hunk 7]
```diff
@@ -1233,6 +1233,166 @@ public void OptionalEmptinessHandlerTest() {
...
+  public void OptionalEmptinessHandlerWithSingleCustomPathTest() {
+    compilationHelper
+        .setArgs(
+            Arrays.asList(
+                "-d",
+                temporaryFolder.getRoot().getAbsolutePath(),
+                "-XepOpt:NullAway:AnnotatedPackages=com.uber",
+                "-XepOpt:NullAway:UnannotatedSubPackages=com.uber.lib.unannotated",
+                "-XepOpt:NullAway:CheckOptionalEmptiness=true",
+                "-XepOpt:NullAway:CheckOptionalEmptinessCustomClasses=com.google.common.base.Optional"))
```
Comment (lazaroclapp) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Please add one more (smaller) test, that checks that `java.util.Optional` is still recognized when this argument is set, as per our discussion. Tests are a good way to document that sort of design edge-case, after all.

Comment (shas19) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: Done already 👍

File: nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java (hunk 6 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: Done 👍

File: nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java [hunk 8]
```diff
@@ -72,14 +72,14 @@ public boolean onOverrideMayBeNullExpr(
...
-                        .map(state.getTypes()::erasure))
-            .collect(Collectors.toSet());
+        ImmutableSet.copyOf(
+            config
+                .getOptionalClassPaths()
+                .stream()
+                .map(state::getTypeFromString)
+                .filter(Objects::nonNull)
+                .map(state.getTypes()::erasure)
+                .collect(Collectors.toSet()));
```
Comment (lazaroclapp) [inline nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: Guava 22 has collectors: https://google.github.io/guava/releases/22.0/api/docs/com/google/common/collect/ImmutableSet.html#toImmutableSet--

Comment (lazaroclapp) [reply nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: So no need to build a Java set and call `copyOf(...)`

File: nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java (hunk 7 above)
Comment (lazaroclapp) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayTest.java]: (True, I missed this one, my bad 😅 )

File: nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java (hunk 8 above)
Comment (shas19) [reply nullaway/src/main/java/com/uber/nullaway/handlers/OptionalEmptinessHandler.java]: Thanks! Done 👍
